### PR TITLE
Added FindOnnx cmake, Removed JSON.cpp & JSON.h copies

### PR DESCRIPTION
**Pull request by Umesh-k26**
- Updated onnx linking with find_package command
- To remove the copies of JSON implementations.
Currently TensorSpec.cpp takes JSON.h from llvm/include if we build it inside LLVM, and from /usr/include if we build it standalone.
Both the files have different implementations for json::Path class, hence TensorSpec.cpp won't compile.
To fix it, compile time flag has been linked with SerDesLib and json::Path has been defined accordingly.

### CODE REVIEW

File: include/MLModelRunner/gRPCModelRunner.h [hunk 1]
```diff
@@ -70,6 +70,7 @@
 #include <grpcpp/grpcpp.h>
 #include <grpcpp/health_check_service_interface.h>
 #include <memory>
+#include <thread>
```
Comment (svkeerthy) [inline include/MLModelRunner/gRPCModelRunner.h:73]: Why is this header required?

File: include/SerDes/baseSerDes.h [hunk 2]
```diff
@@ -43,10 +43,10 @@ namespace MLBridge {
 /// communication by the MLModelRunner.
 /// Currently, (int, float) or (long, double), char and bool are supported.
 /// Vectors of these types are supported as well.
+enum class SerDesKind : int { Unknown, Json, Bitstream, Protobuf, Tensorflow };
```
Comment (svkeerthy) [inline include/SerDes/baseSerDes.h:46]: Why is SerDesKind moved out of BaseSerDes?

File: test/MLBridgeTest.cpp [hunk 3]
```diff
@@ -174,53 +161,67 @@ void increment_port(int delta) {
...
+
   if (server_address == "") {
     std::cerr << "Server Address must be specified via "
                  "--test-server-address=\"<ip>:<port>\"\n";
     exit(1);
   }
   {
     gRPCModelRunnerInit(int);
-    testPrimitive<int, int>("int", 11, 12);
+    testPrimitive<int, int>(MLRunner, "int", 11, 12);
```
Comment (svkeerthy) [inline test/MLBridgeTest.cpp]: Should testPrimitive take three template args?

File: test/MLBridgeTest.cpp [hunk 4]
```diff
@@ -174,53 +161,67 @@ void increment_port(int delta) {
...
   increment_port(1);
   {
     gRPCModelRunnerInit(bool);
-    testPrimitive<bool, bool>("bool", true, false);
+    testPrimitive<bool, bool>(MLRunner, "bool", true, false);
   }
   {
     gRPCModelRunnerInit(vec_int);
-    testVector<int, int>("vec_int", {11, 22, 33}, {12, 23, 34});
+    testVector<int, int>(MLRunner, "vec_int", {11, 22, 33}, {12, 23, 34});
```
Comment (svkeerthy) [inline test/MLBridgeTest.cpp]: testVector should also be taking 3 template args?

File: test/MLBridgeTest.cpp [hunk 5]
```diff
@@ -64,21 +51,21 @@ static llvm::cl::opt<bool>
...
 // send value of type T1. Test received value of type T2 against expected value
-template <typename T1, typename T2>
-void testPrimitive(std::string label, T1 value, T2 expected) {
+template <typename T1, typename T2, typename MLRunnerTy>
+void testPrimitive(MLRunnerTy &MLRunner, std::string label, T1 value,
+                   T2 expected) {
   std::pair<std::string, T1> p("request_" + label, value);
   MLRunner->populateFeatures(p);
-  T2 out = MLRunner->evaluate<T2>();
+  T2 out = MLRunner->template evaluate<T2>();
```
Comment (svkeerthy) [inline test/MLBridgeTest.cpp:68]: What is "template" here? Shouldn't it be the original?

File: test/MLBridgeTest.cpp [hunk 6]
```diff
@@ -87,14 +74,14 @@ void testPrimitive(std::string label, T1 value, T2 expected) {
...
-void testVector(std::string label, std::vector<T1> value,
+template <typename T1, typename T2, typename MLRunnerTy>
+void testVector(MLRunnerTy &MLRunner, std::string label, std::vector<T1> value,
                 std::vector<T2> expected) {
   std::pair<std::string, std::vector<T1>> p("request_" + label, value);
   MLRunner->populateFeatures(p);
   T2 *out;
   size_t size;
-  MLRunner->evaluate<T2 *>(out, size);
+  MLRunner->template evaluate<T2 *>(out, size);
```
Comment (svkeerthy) [inline test/MLBridgeTest.cpp:84]: Same here.

**Review by svkeerthy — CHANGES_REQUESTED** [review]
Please change the README to reflect the upgraded gRPC and protobuf versions.

File: include/MLModelRunner/gRPCModelRunner.h (hunk 1 above)
Comment (Umesh-k26) [reply include/MLModelRunner/gRPCModelRunner.h:73]: gRPCModelRunner uses std::thread to handle graceful exits

File: test/MLBridgeTest.cpp (hunk 5 above)
Comment (Umesh-k26) [reply test/MLBridgeTest.cpp:68]: It is a C++ syntax to tell that member function (evaluate here) is template parameter (T2 here) dependent.

File: test/MLBridgeTest.cpp (hunk 3 above)
Comment (Umesh-k26) [reply test/MLBridgeTest.cpp]: Third one will be inferred from the function arguments.

File: include/SerDes/baseSerDes.h (hunk 2 above)
Comment (svkeerthy) [reply include/SerDes/baseSerDes.h:46]: WIP - More related to CRTP changes. Resolving it for now.

File: test/MLBridgeTest.cpp (hunk 5 above)
Comment (svkeerthy) [reply test/MLBridgeTest.cpp:68]: LG.

**Review by svkeerthy — CHANGES_REQUESTED** [review]

**Review by svkeerthy — APPROVED** [review]

**Pull request merged**